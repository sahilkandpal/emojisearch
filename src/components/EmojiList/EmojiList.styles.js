import Styled from "styled-components";
export const StyledEmojiList = Styled.div`
display: flex;
flex-wrap: wrap;
padding: 20px;
justify-content: center;
@media screen and (max-width: 576px) {
    padding: 0;
    margin-top: 20px;
  }
    .emoji{
        padding: 10px;
        font-size: 30px;
        cursor: pointer;
        width: 40px;
        text-align: center;
        &:hover{
            background-color: #f6f6f6;
        }
        @media screen and (max-width: 576px) {
            width: auto;
          }
    }
    .loading, .notfound {
        text-align: center;
    }   
    .loading{
        font-size: 50px;
        animation: rotate 1s infinite;
    }
    @keyframes rotate {
        0% {transform: rotate(0);}
        50% {transform: rotate(180deg);}
        100% {transform: rotate(0)}
      }
    .notfound{
        &-emoji{
            font-size: 50px;
        }
        &-text{
            font-size: 28px;
            font-weight: 500;
            @media screen and (max-width: 576px) {
                font-size: 24px;
                padding: 10px 0;
              }
        }
    }
`;
