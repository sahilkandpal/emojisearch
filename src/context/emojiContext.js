import { createContext, useState, useContext } from "react";
import emojidata from "../json/emoji-en-US.json";
export const emojiContext = createContext({
  emojiData: {},
  filterEmojiData: () => {},
  handleEmojiData: () => {},
  loading: false,
  handleLoading: () => {},
});
const initialState = emojidata;

export const EmojiContextProvider = ({ children }) => {
  const [emojiData, setEmojiData] = useState(initialState);
  const [loading, setLoading] = useState(false);

  const handleLoading = (state) => {
    setLoading(state);
  };

  const handleEmojiData = (val) => {
    setEmojiData(val);
  };

  const filterEmojiData = (query) => {
    if (query.trim() !== "") {
      let newdata = {};
      Object.keys(emojidata).map((val) => {
        const filtered = emojidata[val].filter((value) =>
          value.toLowerCase().includes(query.trim().toLowerCase())
        );
        if (filtered.length > 0) {
          newdata = { ...newdata, [val]: emojidata[val] };
        }
        return null;
      });
      return newdata;
    } else {
      return emojidata;
    }
  };

  return (
    <emojiContext.Provider
      value={{
        emojiData,
        filterEmojiData,
        handleEmojiData,
        loading,
        handleLoading,
      }}
    >
      {children}
    </emojiContext.Provider>
  );
};

export const useEmojiContext = () => {
  const {
    emojiData,
    filterEmojiData,
    handleEmojiData,
    loading,
    handleLoading,
  } = useContext(emojiContext);
  return {
    emojiData,
    filterEmojiData,
    handleEmojiData,
    loading,
    handleLoading,
  };
};
