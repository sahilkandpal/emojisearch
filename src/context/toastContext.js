import { createContext, useState, useContext } from "react";

const initialState = { message: "", isVisible: false };

export const toastContext = createContext({
  toast: initialState,
  handleToast: () => {},
});

export const ToastContextProvider = ({ children }) => {
  const [toast, setToast] = useState(initialState);

  const handleToast = (message) => {
    setToast({ message, isVisible: true });
    setTimeout(() => {
      setToast({ message, isVisible: false });
    }, 1000);
  };

  return (
    <toastContext.Provider value={{ toast, handleToast }}>
      {children}
    </toastContext.Provider>
  );
};

export const useToastContext = () => {
  const { toast, handleToast } = useContext(toastContext);
  return {
    toast,
    handleToast,
  };
};
