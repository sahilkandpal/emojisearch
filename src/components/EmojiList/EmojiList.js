import React from "react";
import { StyledEmojiList } from "./EmojiList.styles";
import { useEmojiContext } from "../../context/emojiContext";
import { useToastContext } from "../../context/toastContext";
const EmojiList = ({ className }) => {
  const { emojiData, loading } = useEmojiContext();
  const { handleToast } = useToastContext();
  const copyValue = (e) => {
    const emoji = e.target.innerText;
    navigator.clipboard?.writeText(emoji);
    handleToast(`${emoji} copied !`);
  };
  const notFound = Object.keys(emojiData).length === 0;
  const emojis = !loading ? (
    !notFound ? (
      Object.keys(emojiData).map((emoji) => (
        <div className="emoji" key={emoji} onClick={copyValue}>
          {emoji}
        </div>
      ))
    ) : (
      <h1 className="notfound">
        <div className="notfound-emoji">😭</div>
        <div className="notfound-text">No Results</div>
      </h1>
    )
  ) : (
    <h1 className="loading">⏳</h1>
  );

  return <StyledEmojiList className={className}>{emojis}</StyledEmojiList>;
};

export default EmojiList;
