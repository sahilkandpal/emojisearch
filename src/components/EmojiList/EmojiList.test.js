import React from "react";
import { shallow } from "enzyme";
import EmojiList from "./EmojiList";
import { useEmojiContext } from "../../context/emojiContext";

jest.mock("../../context/emojiContext");

describe("Given EmojiList", () => {
  beforeEach(() => {
    useEmojiContext.mockReturnValue({
      emojiData: {},
      loading: false,
    });
  });
  test("when loading is true component should render loading text", () => {
    useEmojiContext.mockReturnValue({
      emojiData: {},
      loading: true,
    });
    const component = shallow(<EmojiList />);
    expect(component).toMatchSnapshot();
  });
  test("when result is not found component should render no result", () => {
    useEmojiContext.mockReturnValue({
      emojiData: {},
      loading: false,
    });
    const component = shallow(<EmojiList />);
    expect(component).toMatchSnapshot();
  });
  test("when result is found component should render emojilist", () => {
    useEmojiContext.mockReturnValue({
      emojiData: {
        "ðŸ˜€": ["grinning_face", "face", "smile", "happy", "joy", ":D", "grin"],
      },
      loading: false,
    });
    const component = shallow(<EmojiList />);
    expect(component).toMatchSnapshot();
  });
});
