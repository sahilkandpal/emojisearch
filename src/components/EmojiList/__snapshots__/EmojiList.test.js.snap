// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given EmojiList when loading is true component should render loading text 1`] = `
<styled.div>
  <h1
    className="loading"
  >
    ⏳
  </h1>
</styled.div>
`;

exports[`Given EmojiList when result is found component should render emojilist 1`] = `
<styled.div>
  <div
    className="emoji"
    key="😀"
    onClick={[Function]}
  >
    😀
  </div>
</styled.div>
`;

exports[`Given EmojiList when result is not found component should render no result 1`] = `
<styled.div>
  <h1
    className="notfound"
  >
    <div
      className="notfound-emoji"
    >
      😭
    </div>
    <div
      className="notfound-text"
    >
      No Results
    </div>
  </h1>
</styled.div>
`;
